### DISCUSSION TITLE
Delete old trajectories on CF

### DISCUSSION THREAD

**Discussion by skovsen** [bitcraze/discussions#1156] [Q&A] [answered]
Hi. I want to upload different compressed trajectories to my CF, all with different lengths. 

I have followed the example in autonomous_sequence_high_level_compressed.py and I use the same trajectory_id for all my uploads.

After some testing I have observed that if I upload a long trajectory, let's say with 10 segments, do the flying, and then upload a trajectory with 9 segments, the last segment of the first trajectory is still active, meaning my CF ends up in the same position as when using the first trajectory.

Is there a way to reset or delete formerly uploaded trajectories?

**Comment by amitx64** [comment]
@skovsen  are you using LPS or Lighthouse or Mocaps?

**Reply by skovsen** [reply]
Hi. I'm using Lighthouse

**Reply by amitx64** [reply]
@skovsen  oh!...

I have LPS and I am trying to do same with LPS....

**Comment by knmcguire** [comment]
Hi! There is no functionality currently that can remove a trajectory. So what happens, is that once a trajectory is uploaded, memory is saved for that trajectory and it stays in there. If you add a new trajectory, it must be the same number of segments to overwrite it fully. In your case, I would just use a different trajectory ID for the 9-segment one, as it will reserve memory for exactly that size so you won't have the problem of the last 10th segment not being removed from memory.

**Reply by skovsen** [reply]
Hi.
Ok, that sounds reasonable. I will try and see if I can work around the problem this way.

I would be very interested in a fix, as you mention below, to release the memory. In the future we are planning on have our CF's fly multiple generated trajectories over a longer period of time, and I can see this becoming a problem in the future, if we have uploaded a very long trajectory, and doesn't get it overwritten properly.

**Comment by knmcguire** [comment]
In a separate discussion along the support question: @whoenig, would it be interesting to have a function that can 'release' the trajectory memory? should we make an issue for this as a wanted functionality?

**Reply by whoenig** [reply]
The allocation of memory for trajectories is fixed in the firmware, so one can "delete" by simply overwriting an existing trajectory. Handling different-length trajectories can be tricky and even if handled properly might lead to fragmentation. The current assumption is that all the memory management would need to be handled on the user application side. A proper support would either require switching to dynamic memory allocation in the firmware, or having a memory manager on the cflib. I personally would avoid both, just because it adds a lot of complexity for an uncommon use-case.

**Reply by skovsen** [reply]
Hi. That makes a lot of sense - I can especially appreciate your reluctance to add alot of complexity for a uncommon case :)

Can you give me any pointers on how to overwrite the entire memory, so I can move forward with using trajectories? I'm having a bit of difficulties finding anything in the documentation.
Thanks!

**Reply by whoenig** [reply, marked as answer]
Currently we have 4k allocated, see https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/crtp_commander_high_level.c#L84.

When you define a trajectory, you provide an offset and a length (through the number of pieces), see https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/crtp_commander_high_level.c#L841-L867. So the easiest might be to double buffer on your client side: use offset 0 for your first large trajectory and offset 2048 for your second large trajectory. Whenever you want to change one, you first write the data of the unused trajectory and then you "define" it (you may reuse the trajectoryId there).

**Reply by skovsen** [reply]
Thanks. I'll look into it.
